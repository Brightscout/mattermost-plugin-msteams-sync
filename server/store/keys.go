package store

func channelsLinkedKey(channelID string) string {
	return "channelsLinked_" + channelID
}

func avatarKey(userID string) string {
	return "avatar_" + userID
}

func mattermostTeamsPostKey(postID string) string {
	return "mattermost_teams_" + postID
}

func teamsMattermostPostKey(postID string) string {
	return "teams_mattermost_" + postID
}

func teamsMattermostChatKey(chatID string) string {
	return "teams_mattermost_chat_" + chatID
}

func tokenForMattermostUserKey(userID string) string {
	return "token_for_mm_user" + userID
}

func tokenForTeamsUserKey(userID string) string {
	return "token_for_teams_user" + userID
}
