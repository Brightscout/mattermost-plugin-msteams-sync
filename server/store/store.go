package store

import (
	"encoding/json"
	"errors"

	"github.com/mattermost/mattermost-server/v6/plugin"
	"golang.org/x/oauth2"
)

const (
	avatarCacheTime = 300
)

type ChannelLink struct {
	MattermostTeam    string
	MattermostChannel string
	MSTeamsTeam       string
	MSTeamsChannel    string
}

type Store interface {
	GetAvatarCache(userID string) ([]byte, error)
	SetAvatarCache(userID string, photo []byte) error
	GetLinkByChannelID(channelID string) (*ChannelLink, error)
	GetLinkByMSTeamsChannelID(teamID, channelID string) (*ChannelLink, error)
	DeleteLinkByChannelID(channelID string) error
	StoreChannelLink(link *ChannelLink) error
	TeamsToMattermostPostId(chatID string, postID string) (string, error)
	MattermostToTeamsPostId(postID string) (string, error)
	LinkPosts(mattermostPostID, chatOrChannelID, teamsPostID string) error
	GetTokenForMattermostUser(userID string) (*oauth2.Token, error)
	SetTokenForMattermostUser(userID string, token *oauth2.Token) error
	SetTeamsToMattermostUserId(teamsUserID, mattermostUserId string) error
	SetMattermostToTeamsUserId(mattermostUserId, teamsUserId string) error
	TeamsToMattermostUserId(userID string) (string, error)
	MattermostToTeamsUserId(userID string) (string, error)
	CheckEnabledTeamByTeamId(teamId string) bool
}

type StoreImpl struct {
	api          plugin.API
	enabledTeams func() []string
}

func New(api plugin.API, enabledTeams func() []string) *StoreImpl {
	return &StoreImpl{
		api:          api,
		enabledTeams: enabledTeams,
	}
}

func (s *StoreImpl) GetAvatarCache(userID string) ([]byte, error) {
	data, appErr := s.api.KVGet(avatarKey(userID))
	if appErr != nil {
		return nil, appErr
	}
	return data, nil
}

func (s *StoreImpl) SetAvatarCache(userID string, photo []byte) error {
	appErr := s.api.KVSetWithExpiry(avatarKey(userID), photo, avatarCacheTime)
	if appErr != nil {
		return appErr
	}
	return nil

}

func (s *StoreImpl) GetLinkByChannelID(channelID string) (*ChannelLink, error) {
	linkdata, appErr := s.api.KVGet(channelsLinkedKey(channelID))
	if appErr != nil {
		return nil, appErr
	}
	var link ChannelLink
	err := json.Unmarshal(linkdata, &link)
	if err != nil {
		return nil, err
	}
	if !s.CheckEnabledTeamByTeamId(link.MattermostTeam) {
		return nil, errors.New("link not enabled for this team")
	}
	return &link, nil
}

func (s *StoreImpl) GetLinkByMSTeamsChannelID(teamID, channelID string) (*ChannelLink, error) {
	linkdata, appErr := s.api.KVGet(channelsLinkedByMSTeamsKey(teamID, channelID))
	if appErr != nil {
		return nil, appErr
	}
	var link ChannelLink
	err := json.Unmarshal(linkdata, &link)
	if err != nil {
		return nil, err
	}
	if !s.CheckEnabledTeamByTeamId(link.MattermostTeam) {
		return nil, errors.New("link not enabled for this team")
	}
	return &link, nil
}

func (s *StoreImpl) DeleteLinkByChannelID(channelID string) error {
	_, err := s.GetLinkByChannelID(channelID)
	if err != nil {
		return err
	}

	appErr := s.api.KVDelete(channelsLinkedKey(channelID))
	if appErr != nil {
		return appErr
	}

	return nil
}

func (s *StoreImpl) StoreChannelLink(link *ChannelLink) error {
	linkdata, err := json.Marshal(link)
	if err != nil {
		return err
	}

	appErr := s.api.KVSet(channelsLinkedKey(link.MattermostChannel), linkdata)
	if appErr != nil {
		return appErr
	}
	appErr = s.api.KVSet(channelsLinkedByMSTeamsKey(link.MSTeamsTeam, link.MSTeamsChannel), linkdata)
	if appErr != nil {
		_ = s.api.KVDelete(channelsLinkedKey(link.MattermostChannel))
		return appErr
	}
	return nil
}

func (s *StoreImpl) TeamsToMattermostUserId(userID string) (string, error) {
	data, err := s.api.KVGet(teamsMattermostUserKey(userID))
	if err != nil {
		return "", err
	}
	return string(data), nil
}

func (s *StoreImpl) MattermostToTeamsUserId(userID string) (string, error) {
	data, err := s.api.KVGet(mattermostTeamsUserKey(userID))
	if err != nil {
		return "", err
	}
	return string(data), nil
}

func (s *StoreImpl) TeamsToMattermostPostId(chatID string, postID string) (string, error) {
	data, err := s.api.KVGet(teamsMattermostPostKey(chatID, postID))
	if err != nil {
		return "", err
	}
	return string(data), nil
}

func (s *StoreImpl) MattermostToTeamsPostId(postID string) (string, error) {
	data, err := s.api.KVGet(mattermostTeamsPostKey(postID))
	if err != nil {
		return "", err
	}
	return string(data), nil
}

func (s *StoreImpl) LinkPosts(mattermostPostID, teamsChatOrChannelID, teamsPostID string) error {
	err := s.api.KVSet(mattermostTeamsPostKey(mattermostPostID), []byte(teamsPostID))
	if err != nil {
		return err
	}
	err = s.api.KVSet(teamsMattermostPostKey(teamsChatOrChannelID, teamsPostID), []byte(mattermostPostID))
	if err != nil {
		_ = s.api.KVDelete(mattermostTeamsPostKey(mattermostPostID))
		return err
	}
	return nil
}

func (s *StoreImpl) GetTokenForMattermostUser(userID string) (*oauth2.Token, error) {
	tokendata, appErr := s.api.KVGet(tokenForMattermostUserKey(userID))
	if appErr != nil {
		return nil, appErr
	}
	var token oauth2.Token
	err := json.Unmarshal(tokendata, &token)
	if err != nil {
		return nil, err
	}
	return &token, nil
}

func (s *StoreImpl) SetTokenForMattermostUser(userID string, token *oauth2.Token) error {
	tokendata, err := json.Marshal(token)
	if err != nil {
		return err
	}
	appErr := s.api.KVSet(tokenForMattermostUserKey(userID), tokendata)
	if appErr != nil {
		return appErr
	}
	return nil
}

func (s *StoreImpl) SetTeamsToMattermostUserId(teamsUserID, mattermostUserId string) error {
	appErr := s.api.KVSet(teamsMattermostUserKey(teamsUserID), []byte(mattermostUserId))
	if appErr != nil {
		return appErr
	}
	return nil
}

func (s *StoreImpl) SetMattermostToTeamsUserId(mattermostUserId, teamsUserID string) error {
	appErr := s.api.KVSet(mattermostTeamsUserKey(mattermostUserId), []byte(teamsUserID))
	if appErr != nil {
		return appErr
	}
	return nil
}

func (s *StoreImpl) CheckEnabledTeamByTeamId(teamId string) bool {
	if len(s.enabledTeams()) == 1 && s.enabledTeams()[0] == "" {
		return true
	}
	team, appErr := s.api.GetTeam(teamId)
	if appErr != nil {
		return false
	}
	isTeamEnabled := false
	for _, enabledTeam := range s.enabledTeams() {
		if team.Name == enabledTeam {
			isTeamEnabled = true
			break
		}
	}
	return isTeamEnabled
}
