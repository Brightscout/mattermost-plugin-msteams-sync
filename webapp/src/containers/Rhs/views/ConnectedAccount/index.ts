export {ConnectedAccount} from './ConnectedAccount.container';
