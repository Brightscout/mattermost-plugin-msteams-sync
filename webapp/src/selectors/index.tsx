import {ApiRequestCompletionState, ConnectedState, ReduxState, SnackbarState} from 'types/common/store.d';

export const getApiRequestCompletionState = (state: ReduxState['plugins-com.mattermost.msteams-sync']): ApiRequestCompletionState => state.apiRequestCompletionSlice;

export const getConnectedState = (state: ReduxState['plugins-com.mattermost.msteams-sync']): ConnectedState => state.connectedStateSlice;

export const getSnackbarState = (state: ReduxState['plugins-com.mattermost.msteams-sync']): SnackbarState => state.snackbarSlice;

export const getIsRhsLoading = (state: ReduxState['plugins-com.mattermost.msteams-sync']): {isRhsLoading: boolean} => state.rhsLoadingSlice;
