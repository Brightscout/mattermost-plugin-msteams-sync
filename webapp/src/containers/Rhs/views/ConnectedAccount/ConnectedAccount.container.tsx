import React, {useCallback, useEffect, useState} from 'react';
import {useDispatch} from 'react-redux';

import {Button} from '@brightscout/mattermost-ui-library';

import {pluginApiServiceConfigs} from 'constants/apiService.constant';

import {Dialog} from 'components';

import usePluginApi from 'hooks/usePluginApi';
import useAlert from 'hooks/useAlert';
import useApiRequestCompletionState from 'hooks/useApiRequestCompletionState';

import {setConnected} from 'reducers/connectedState';

import {getConnectedState} from 'selectors';

import utils from 'utils';

export const ConnectedAccount = () => {
    const dispatch = useDispatch();
    const [showDialog, setShowDialog] = useState(false);
    const {makeApiRequestWithCompletionStatus, state} = usePluginApi();
    const {username, isAlreadyConnected, msteamsUserId, connected} = getConnectedState(state);

    const showAlert = useAlert();

    const disconnectUser = useCallback(() => {
        makeApiRequestWithCompletionStatus(pluginApiServiceConfigs.disconnectUser.apiServiceName);
    }, []);

    useApiRequestCompletionState({
        serviceName: pluginApiServiceConfigs.disconnectUser.apiServiceName,
        handleSuccess: () => {
            dispatch(setConnected({connected: false, username: '', msteamsUserId: '', isAlreadyConnected: false}));
            setShowDialog(false);
            showAlert({
                message: 'Your account has been disconnected.',
            });
        },
        handleError: () => {
            showAlert({
                message: 'Error occurred while disconnecting the user.',
                severity: 'error',
            });
            setShowDialog(false);
        },
    });

    useEffect(() => {
        if (connected && !isAlreadyConnected) {
            showAlert({message: 'Your account is connected successfully.'});
            dispatch(setConnected({connected, msteamsUserId, username, isAlreadyConnected: true}));
        }
    }, [connected, isAlreadyConnected]);

    return (
        <div className='msteams-sync-utils'>
            <div className='flex-1 msteams-sync-rhs d-flex flex-column'>
                <div className='py-12 px-20 border-y-1 d-flex gap-8'>
                    {/* TODO: Refactor user Avatar */}
                    <div
                        style={{
                            height: '32px',
                            width: '32px',
                            borderRadius: '50%',
                            backgroundColor: 'rgba(var(--center-channel-color-rgb), 0.12)',
                        }}
                    >
                        <img
                            style={{
                                borderRadius: '50%',
                            }}
                            src={utils.getAvatarUrl(msteamsUserId)}
                        />
                    </div>
                    <div>
                        <h5 className='my-0 font-12 lh-16'>{'Connected as '}<span className='wt-600'>{username}</span></h5>
                        <Button
                            size='sm'
                            variant='text'
                            className='p-0 lh-16'
                            onClick={() => setShowDialog(true)}
                        >{'Disconnect'}</Button>
                    </div>
                </div>
                <div className='d-flex align-items-center justify-center flex-1 flex-column px-40'>
                    {/* NOTE: Part of Phase-II */}
                    {/* <Icon iconName='noChannels'/>
                <h3 className='my-0 lh-28 wt-600 text-center'>{'There are no linked channels yet'}</h3> */}
                </div>
                <Dialog
                    show={showDialog}
                    destructive={true}
                    primaryButtonText='Disconnect'
                    secondaryButtonText='Cancel'
                    isLoading={false}
                    title='Disconnect Microsoft Teams Account'
                    onCloseHandler={() => setShowDialog(false)}
                    onSubmitHandler={disconnectUser}
                >
                    {'Are you sure you want to disconnect your Microsoft Teams Account? You will no longer be able to send and receive messages to Microsoft Teams users from Mattermost.'}
                </Dialog>
            </div>
        </div>
    );
};
