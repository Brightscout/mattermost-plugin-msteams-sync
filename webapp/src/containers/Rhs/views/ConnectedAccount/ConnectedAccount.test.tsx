import React from 'react';

import {RenderResult, render} from '@testing-library/react';
import userEvent from '@testing-library/user-event';

import {getConnectedState} from 'selectors';

import {mockDispatch} from 'tests/setup';

import {mockTestState} from 'tests/mockState';

import {ReduxState} from 'types/common/store.d';

import {ConnectedAccount} from './ConnectedAccount.container';

let tree: RenderResult;

describe('Connected Account View', () => {
    beforeEach(() => {
        tree = render(<ConnectedAccount/>);
    });

    it('should render correctly', () => {
        expect(tree).toMatchSnapshot();
    });

    it('should render disconnect account button', () => {
        const disconnectButton = tree.getByText('Disconnect');
        expect(disconnectButton).toBeVisible();
    });

    it('should display the name of connected user correctly', () => {
        const usernameContainer = tree.getByText('Connected as');
        const username = tree.getByText(getConnectedState(mockTestState as ReduxState).username);

        expect(usernameContainer).toContainElement(username);
        expect(username).toBeVisible();
    });

    it('should dispatch action on clicking disconnect', async () => {
        const disconnectButton = tree.getByText('Disconnect');

        await userEvent.click(disconnectButton);
        expect(mockDispatch).toHaveBeenCalledTimes(1);
    });
});
