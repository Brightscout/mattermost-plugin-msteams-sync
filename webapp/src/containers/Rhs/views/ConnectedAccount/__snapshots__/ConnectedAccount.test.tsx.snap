// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Connected Account View Should render correctly 1`] = `
{
  "asFragment": [Function],
  "baseElement": <body>
    <div>
      <div
        class="msteams-sync-utils"
      >
        <div
          class="flex-1 msteams-sync-rhs d-flex flex-column"
        >
          <div
            class="py-12 px-20 border-y-1 d-flex gap-8"
          >
            <div
              style="height: 32px; width: 32px; border-radius: 50%;"
            >
              <img
                src="http://localhost/plugins/com.mattermost.msteams-sync/avatar/mock_msteams_user_id"
                style="border-radius: 50%;"
              />
            </div>
            <div>
              <h5
                class="my-0 font-12 lh-16"
              >
                Connected as 
                <span
                  class="wt-600"
                >
                  John Doe
                </span>
              </h5>
              <button
                class="sc-kEqXSa fSZMRH mm-button p-0 lh-16"
                width="fit-content"
              >
                Disconnect
              </button>
            </div>
          </div>
          <div
            class="d-flex align-items-center justify-center flex-1 flex-column px-40"
          />
        </div>
      </div>
    </div>
  </body>,
  "container": <div>
    <div
      class="msteams-sync-utils"
    >
      <div
        class="flex-1 msteams-sync-rhs d-flex flex-column"
      >
        <div
          class="py-12 px-20 border-y-1 d-flex gap-8"
        >
          <div
            style="height: 32px; width: 32px; border-radius: 50%;"
          >
            <img
              src="http://localhost/plugins/com.mattermost.msteams-sync/avatar/mock_msteams_user_id"
              style="border-radius: 50%;"
            />
          </div>
          <div>
            <h5
              class="my-0 font-12 lh-16"
            >
              Connected as 
              <span
                class="wt-600"
              >
                John Doe
              </span>
            </h5>
            <button
              class="sc-kEqXSa fSZMRH mm-button p-0 lh-16"
              width="fit-content"
            >
              Disconnect
            </button>
          </div>
        </div>
        <div
          class="d-flex align-items-center justify-center flex-1 flex-column px-40"
        />
      </div>
    </div>
  </div>,
  "debug": [Function],
  "findAllByAltText": [Function],
  "findAllByDisplayValue": [Function],
  "findAllByLabelText": [Function],
  "findAllByPlaceholderText": [Function],
  "findAllByRole": [Function],
  "findAllByTestId": [Function],
  "findAllByText": [Function],
  "findAllByTitle": [Function],
  "findByAltText": [Function],
  "findByDisplayValue": [Function],
  "findByLabelText": [Function],
  "findByPlaceholderText": [Function],
  "findByRole": [Function],
  "findByTestId": [Function],
  "findByText": [Function],
  "findByTitle": [Function],
  "getAllByAltText": [Function],
  "getAllByDisplayValue": [Function],
  "getAllByLabelText": [Function],
  "getAllByPlaceholderText": [Function],
  "getAllByRole": [Function],
  "getAllByTestId": [Function],
  "getAllByText": [Function],
  "getAllByTitle": [Function],
  "getByAltText": [Function],
  "getByDisplayValue": [Function],
  "getByLabelText": [Function],
  "getByPlaceholderText": [Function],
  "getByRole": [Function],
  "getByTestId": [Function],
  "getByText": [Function],
  "getByTitle": [Function],
  "queryAllByAltText": [Function],
  "queryAllByDisplayValue": [Function],
  "queryAllByLabelText": [Function],
  "queryAllByPlaceholderText": [Function],
  "queryAllByRole": [Function],
  "queryAllByTestId": [Function],
  "queryAllByText": [Function],
  "queryAllByTitle": [Function],
  "queryByAltText": [Function],
  "queryByDisplayValue": [Function],
  "queryByLabelText": [Function],
  "queryByPlaceholderText": [Function],
  "queryByRole": [Function],
  "queryByTestId": [Function],
  "queryByText": [Function],
  "queryByTitle": [Function],
  "rerender": [Function],
  "unmount": [Function],
}
`;
